phosphor-auditlog/alog_parser.hpp

```diff
@@ -0,0 +1,258 @@
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatMsgReg(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Formats next record into JSON general format
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatGen(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Formats next record into JSON raw format
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatRaw(nlohmann::json& parsedEntry);
+
+    auparse_state_t* au = nullptr;
+    std::ofstream parsedStream;
+
+  private:
+    /**
+     * @brief Moves parser to point to next event
+     *
+     * @return false when no more events exist, or on error
+     */
+    bool auditNextEvent();
+
+    /**
+     * @brief Parses next event and each of its records into JSON format
+     * @details Writes the audit log events to parsedStream.
+     */
+    void parseEvent();
+
+    /**
+     * @brief Parses and writes next record into JSON format
+     */
+    virtual void parseRecord();
+
+    /**
+     * @brief Parses general audit entry into JSON format
+     * @details Used with audit entries without specific handling. Text of audit
+     * log message is written as-is.
+     */
+    void fillAuditEntry(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Parses AUDIT_USYS_CONFIG audit entry into JSON format
+     * @details Expected fields from audit log entry are split into MessageArgs
+     */
+    void fillUsysEntry(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Opens and truncates specified file
+     * @param[in] filePath Path of file to open. File should exist.
+     * @return bool True if stream was established to file, false otherwise.
+     */
+    bool openParsedFile(const std::string& filePath);
+};
+
+/** @class ALParseLatest
+ *  @brief Parsing audit log using auparse library services
+ *  @details Provides means to parse only latest entries
+ */
+class ALParseLatest : public ALParser
+{
+  public:
+    ALParseLatest(const ALParseLatest&) = delete;
+    ALParseLatest& operator=(const ALParseLatest&) = delete;
+    ALParseLatest(ALParseLatest&&) = delete;
+    ALParseLatest& operator=(ALParseLatest&&) = delete;
+
+    /** @brief Constructor to initialize parsing of audit log files
+     *  @param[in] maxEvents Maximum number of events to return. Minimum
+     *             allowed value is 1.
+     *  @param[in] parsedFile Initialized file for holding parsed log events
+     */
+    ALParseLatest(uint32_t maxEvents, ALParseFile& parsedFile) :
+        ALParser(parsedFile)
+    {
+        lg2::debug("Constructing ALParseLatest: {MAXCOUNT}", "MAXCOUNT",
+                   maxEvents);
+
+        // Don't allow request without non-zero count
+        if (maxEvents > 0)
```

Comment:
if the D-Bus call was made with a maxEvents of zero, would it get to line 170 and then return an InternalFailure?  Ideally that would fail with InvalidArgument (or just return zero entries back)

Reply:
Yes it was returning InternalFailure. I agree that wasn't the best approach.

I've decided to just return an empty file if maxEvents is zero. I'll need to update the YAML file description since it specified minimum value accepted was 1. But I think it is better to handle the case then report an error.